### PR TITLE
Add cd74hc4067 multiplexer

### LINKED ISSUE #85: [Feature Request] Support CD74HC4067 16 Channel Analog / Digital Multiplexer...
**Describe the problem you have/What new integration you would like**

It would be great to be able to add the multiplexer to Analog or digital and then have more ports to use for integration. By doing this we can even use D1 Mini with expander boards to be able to expand even more.  This can bring the count down of connected devices on your network.

**Please describe your use case for this integration:**

I want to use this for alarm port monitoring (Digital for new alarm system only using This, or connecting to an analog port for piggy back on Alarm systems with resistors)

With homes that you want to use presence detection, many homes have large installations of Alarm systems and we can use the PIR, Door Magnets to do automation with.  Why add more devices if you can use existing once.

I have an IDS x64 that can do 64x zones and unable to connect directly to it, so piggy back is the only way.

**Additional context**

Here is an Arduino Library for this;
https://github.com/waspinator/CD74HC4067

Thank you,
Andrew

### PR DESCRIPTION
**Pull request by asoehlke**
# What does this implement/fix? 

Addition of component/sensor for CD74HC4067 analog multiplexer

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes esphome/feature-requests#85

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1503

## Test Environment

- [ ] ESP32
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
cd74hc4067:
  pin_s0: D0
  pin_s1: D1
  pin_s2: D2
  pin_s3: D3     

sensor:
  - platform: adc
    id: adc_sensor
    pin: A0
    update_interval: 3600s
  - platform: cd74hc4067
    id: solar_1_raw
    number: 0
    sensor: adc_sensor
    update_interval: 3600s
  - platform: cd74hc4067
    id: solar_2_raw
    number: 1
    sensor: adc_sensor
    update_interval: 3600s
  - platform: ct_clamp
    name: "SolarPower1"
    sensor: solar_1_raw
    update_interval: 5s
    unit_of_measurement: "A"
    filters:
        - multiply: 36000
  - platform: ct_clamp
    name: "SolarPower2"
    sensor: solar_2_raw
    update_interval: 5s
    unit_of_measurement: "A"
    filters:
        - multiply: 36000
      
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/cd74hc4067/cd74hc4067.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
+#include "cd74hc4067.h"
+#include "esphome/core/log.h"
+#include <Arduino.h>
```
Comment (OttoWinter) [inline esphome/components/cd74hc4067/cd74hc4067.cpp]: This include won't work on esp-idf based builds; I don't see it used anywhere, so let's remove it

File: esphome/components/cd74hc4067/sensor.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+DEPENDENCIES = ["cd74hc4067"]
+
+CD74HC4067Sensor = cd74hc4067_ns.class_(
+    "CD74HC4067Sensor",
+    sensor.Sensor,
+    cg.PollingComponent,
+    voltage_sampler.VoltageSampler,
+)
+
+CONF_cd74hc4067_ID = "cd74hc4067_id"
```
Comment (OttoWinter) [inline esphome/components/cd74hc4067/sensor.py]:
```suggestion
CONF_CD74HC4067_ID = "cd74hc4067_id"
```

Those CONF_ names are always uppercase in esphome

Comment (asoehlke) [reply esphome/components/cd74hc4067/sensor.py]: changed

File: esphome/components/cd74hc4067/cd74hc4067.cpp (hunk 1 above)
Comment (asoehlke) [reply esphome/components/cd74hc4067/cd74hc4067.cpp]: Thanks for the utra fast review, the import is removed. 
Was a remainder of accessing the pin directly.

**Comment by asoehlke** [comment]
Sorry, false alert for code owner @jeromelaban, @ncareau, @Syndlex, @OttoWinter, @trvrnrth, @esphome/core, the bot was faster than me cleaning up some mess in this branch. Your components are not changed!

**Comment by ekorsanov01** [comment]
When will this component be added to the master?

**Comment by asoehlke** [comment]
> When will this component be added to the master?

For me: anytime :-)
A second review from @jesserockz is still pending and then I hope one of the maintainers can merge it.
If you like you can also clone from my fork (https://github.com/asoehlke/esphome) and try it out.

**Comment by csk06** [comment]
@asoehlke 

Hi Andreas, 

I need to use this unit in my project, so for that it seemed to change the esphome_version to your repo. But it seems like esphome has removed that esphome_version from configuration page. 

Can you please tell me what i need to do in order to "integrate" your codes to my setup. 

Thanks

**Comment by jesserockz** [comment]
@csk06 Until this is merged you can get it in by using this then the documented method for the component:

```yaml
external_components:
  - source: github://pr#2431
    components: ["cd74hc4067"]
```

File: esphome/components/cd74hc4067/cd74hc4067.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    delay(switch_delay_);
+    active_pin_ = pin;
+  }
+}
+
+CD74HC4067Sensor::CD74HC4067Sensor(CD74HC4067Component *parent, const std::string &name, uint8_t pin)
+    : PollingComponent(1000), parent_(parent), pin_(pin) {
+  this->set_name(name);
+}
+void CD74HC4067Sensor::setup() { LOG_SENSOR("", "Setting up CD74HC4067 Multiplexer Channel", this); }
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/cd74hc4067.cpp]: No need for a whole setup function that just sends a log message...

File: esphome/components/cd74hc4067/cd74hc4067.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    // small delay is needed to let the multiplexer switch
+    delay(switch_delay_);
+    active_pin_ = pin;
+  }
+}
+
+CD74HC4067Sensor::CD74HC4067Sensor(CD74HC4067Component *parent, const std::string &name, uint8_t pin)
+    : PollingComponent(1000), parent_(parent), pin_(pin) {
+  this->set_name(name);
+}
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/cd74hc4067.cpp]: Can remove all references to `name` as it is set by the internal EntityBase code

File: esphome/components/cd74hc4067/cd74hc4067.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+void CD74HC4067Sensor::setup() { LOG_SENSOR("", "Setting up CD74HC4067 Multiplexer Channel", this); }
+
+void CD74HC4067Sensor::update() {
+  float value_v = this->sample();
+  this->publish_state(value_v);
+}
+
+float CD74HC4067Sensor::get_setup_priority() const { return parent_->get_setup_priority() - 1.0f; }
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/cd74hc4067.cpp]: Consistency with use of `this->` please
```suggestion
float CD74HC4067Sensor::get_setup_priority() const { this->return parent_->get_setup_priority() - 1.0f; }
```

File: esphome/components/cd74hc4067/cd74hc4067.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+float CD74HC4067Sensor::sample() {
+  this->parent_->activate_pin(pin_);
+  return source_->sample();
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/cd74hc4067.cpp]:
```suggestion
  this->parent_->activate_pin(this->pin_);
  return this->source_->sample();
```

File: esphome/components/cd74hc4067/cd74hc4067.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+std::string CD74HC4067Sensor::unique_id() { return get_mac_address() + "-" + to_string(pin_); }
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/cd74hc4067.cpp]: This should at least match the format of the default unique id please, but really shouldn't need overriding at all to keep the code cleaner.

```
std::string get_default_unique_id(const std::string &component_type, EntityBase *entity) {
  return App.get_name() + component_type + entity->get_object_id();
}
```

File: esphome/components/cd74hc4067/cd74hc4067.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+void CD74HC4067Sensor::dump_config() {
+  LOG_SENSOR(TAG, "Multiplexer Sensor", this);
+  ESP_LOGCONFIG(TAG, "CD74HC4067 Pin: %u", this->pin_);
+  LOG_UPDATE_INTERVAL(this);
+}
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/cd74hc4067.cpp]:
```suggestion
void CD74HC4067Sensor::dump_config() {
  LOG_SENSOR(TAG, "CD74HC4067 Sensor", this);
  ESP_LOGCONFIG(TAG, "  Pin: %u", this->pin_);
  LOG_UPDATE_INTERVAL(this);
}
```

File: esphome/components/cd74hc4067/cd74hc4067.h [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  InternalGPIOPin *pin_s2_;
+  InternalGPIOPin *pin_s3_;
+  /// the currently active pin
+  uint8_t active_pin_;
+  uint32_t switch_delay_;
+};
+
+class CD74HC4067Sensor : public sensor::Sensor, public PollingComponent, public voltage_sampler::VoltageSampler {
+ public:
+  CD74HC4067Sensor(CD74HC4067Component *parent, const std::string &name, uint8_t pin);
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/cd74hc4067.h]: Please remove the name and pin from the constructor and use the `set_pin` setter function below from `to_code` too.

Could also remove the constructor definition altogether and extend `public Parented<CD74HC4067Component>` too and use `await cg.register_parented(var, config[CONF_CD74HC4067_ID])`

```suggestion
  CD74HC4067Sensor(CD74HC4067Component *parent);
```

File: esphome/components/cd74hc4067/sensor.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    )
+    .extend(cv.polling_component_schema("60s"))
+)
+
+
+async def to_code(config):
+    parent = await cg.get_variable(config[CONF_CD74HC4067_ID])
+    var = cg.new_Pvariable(
+        config[CONF_ID], parent, config[CONF_NAME], config[CONF_NUMBER]
+    )
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/sensor.py]: As per above
```suggestion
    var = cg.new_Pvariable(config[CONF_ID], parent)
```

File: esphome/components/cd74hc4067/sensor.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    voltage_sampler.VoltageSampler,
+)
+
+CONF_CD74HC4067_ID = "cd74hc4067_id"
+
+CONFIG_SCHEMA = (
+    sensor.sensor_schema(
+        unit_of_measurement=UNIT_WATT,
+        accuracy_decimals=3,
+        device_class=DEVICE_CLASS_POWER,
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/sensor.py]: Shouldn't this be a `UNIT_VOLT` and `DEVICE_CLASS_VOLTAGE` by default?

File: esphome/components/cd74hc4067/sensor.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+async def to_code(config):
+    parent = await cg.get_variable(config[CONF_CD74HC4067_ID])
+    var = cg.new_Pvariable(
+        config[CONF_ID], parent, config[CONF_NAME], config[CONF_NUMBER]
+    )
+    await cg.register_component(var, config)
+    await sensor.register_sensor(var, config)
+
+    sens = await cg.get_variable(config[CONF_SENSOR])
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/sensor.py:54]:
```suggestion
    cg.add(var.set_pin(config[CONF_NUMBER]))

    sens = await cg.get_variable(config[CONF_SENSOR])
```

File: tests/test3.yaml [hunk 13]
```diff
@@ -1339,3 +1343,9 @@ dsmr:
 daly_bms:
   update_interval: 20s
   uart_id: uart1
+
+cd74hc4067:
+  pin_s0: D0
+  pin_s1: D1
+  pin_s2: D2
+  pin_s3: D3
```
Comment (jesserockz) [inline tests/test3.yaml]: Not sure about the rest of the file, but please use GPIOx pins not Dx pins as they are board specific.

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/cd74hc4067/cd74hc4067.cpp (hunk 3 above)
Comment (asoehlke) [reply esphome/components/cd74hc4067/cd74hc4067.cpp]: Removed

File: esphome/components/cd74hc4067/cd74hc4067.cpp (hunk 4 above)
Comment (asoehlke) [reply esphome/components/cd74hc4067/cd74hc4067.cpp]: Name handling removed

File: esphome/components/cd74hc4067/cd74hc4067.cpp (hunk 5 above)
Comment (asoehlke) [reply esphome/components/cd74hc4067/cd74hc4067.cpp]: corrected, I hope I got all places.

File: esphome/components/cd74hc4067/cd74hc4067.cpp (hunk 6 above)
Comment (asoehlke) [reply esphome/components/cd74hc4067/cd74hc4067.cpp]: corrected

File: esphome/components/cd74hc4067/cd74hc4067.cpp (hunk 7 above)
Comment (asoehlke) [reply esphome/components/cd74hc4067/cd74hc4067.cpp]: removed, should really not be necessary with the name set correctly.

File: esphome/components/cd74hc4067/sensor.py (hunk 11 above)
Comment (asoehlke) [reply esphome/components/cd74hc4067/sensor.py]: Yes, these are better defaults, changed.

File: esphome/components/cd74hc4067/cd74hc4067.h (hunk 9 above)
Comment (asoehlke) [reply esphome/components/cd74hc4067/cd74hc4067.h]: changed

File: esphome/components/cd74hc4067/sensor.py (hunk 12 above)
Comment (asoehlke) [reply esphome/components/cd74hc4067/sensor.py:54]: changed

File: tests/test3.yaml (hunk 13 above)
Comment (asoehlke) [reply tests/test3.yaml]: changed

**Comment by asoehlke** [comment]
Updated the device with the changed version, looks good, but I would like to wait for the sun to rise here (Berlin time) to check whether the values are as expected.

File: esphome/components/cd74hc4067/cd74hc4067.cpp [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    this->pin_s1_->digital_write(mux_channel[pin][1]);
+    this->pin_s2_->digital_write(mux_channel[pin][2]);
+    this->pin_s3_->digital_write(mux_channel[pin][3]);
+    // small delay is needed to let the multiplexer switch
+    delay(this->switch_delay_);
+    this->active_pin_ = pin;
+  }
+}
+
+CD74HC4067Sensor::CD74HC4067Sensor(CD74HC4067Component *parent) : PollingComponent(1000), parent_(parent) {}
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/cd74hc4067.cpp]: I missed this before, not needed and will be set by the internal code.
```suggestion
CD74HC4067Sensor::CD74HC4067Sensor(CD74HC4067Component *parent) : parent_(parent) {}
```

File: esphome/components/cd74hc4067/cd74hc4067.cpp [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+void CD74HC4067Sensor::update() {
+  float value_v = this->sample();
+  this->publish_state(value_v);
+}
+
+float CD74HC4067Sensor::get_setup_priority() const { return this->parent_->get_setup_priority() - 1.0f; }
+
+float CD74HC4067Sensor::sample() {
+  this->parent_->activate_pin(pin_);
```
Comment (jesserockz) [inline esphome/components/cd74hc4067/cd74hc4067.cpp]:
```suggestion
  this->parent_->activate_pin(this->pin_);
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**